UndertaleModLib/Decompiler/Decompiler.cs

```diff
@@ -268,20 +267,16 @@ public bool EqualsNumber(int TestNumber)
                 {
                     return Convert.ToInt32(val);
                 }
-                else if (val is double)
-                {
-                    var v = Convert.ToDouble(val);
-                    int res = (int)v;
-                    if (v == res)
-                        return res;
-                }
-                else if (val is float)
+
+                else if (val is double || val is float)
                 {
-                    var v = Convert.ToSingle(val);
-                    int res = (int)v;
-                    if (v == res)
-                        return res;
+                    // Handle float and doubles properly
+                    var Converted = (val is double) ? Convert.ToDouble(val) : Convert.ToSingle(val);
```

Comment:
I think this might actually result in a performance degradation as compared to the old method, since it not only requires a conversion to double for floats, but it also checks the value's type at least twice, and then casts to int twice (though that is probably easily optimized by the compiler, at least).